src/gui/render/graphics_shaders/program_handler.cpp

```diff
@@ -0,0 +1,340 @@
+#include "program_handler.hpp"
+
+#include "gui_render_types.hpp"
+#include "logging.hpp"
+
+#include <GL/glew.h>
+
+#include <filesystem>
+#include <fstream>
+#include <regex>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace gui {
+
+namespace shader {
+
+std::optional<std::string>
+File::get_file_content() {
+    std::ifstream shader_stream(file_, std::ios::in);
+    if (shader_stream.is_open()) {
+        std::stringstream sstr;
+        sstr << shader_stream.rdbuf();
+        std::string shader_code = sstr.str();
+        shader_stream.close();
+        status_ = FileStatus::OK;
+        return shader_code;
+    } else {
+        LOG_ERROR(
+            logging::opengl_logger, "Cannot open {}. Is this the right directory?",
+            file_.string()
+        );
+        status_ = FileStatus::FILE_NOT_FOUND;
+        return {};
+    }
+}
+
+std::string
+get_shader_string(GLuint gl_shader_type) {
+    switch (gl_shader_type) {
+        case GL_VERTEX_SHADER:
+            return "vertex";
+        case GL_FRAGMENT_SHADER:
+            return "fragment";
+        case GL_GEOMETRY_SHADER:
+            return "geometry";
+
+        default:
+            return "NOT A VALID SHADER TYPE";
```

Comment:
```suggestion
            abort();
```

Reply:
You're switching based on an enum. You cover every value of that enum. If the value passed to the function somehow isn't in that enum, things have gone terribly wrong, and you should exit as quickly as possible.

Reply:
GLuint is not an enum? This is literally a functions that retuns a string. If one randomly puts a number in here why should it crash?

Reply:
> GLuint is not an enum?

You can make an `enum` out of `#define`s.

> GLuint is not an enum? This is literally a functions that retuns a string. If one randomly puts a number in here why should it crash?

Because, unless a cosmic ray hits your DRAM, or you have a memory corruption bug, the `default` case should never run. And if a cosmic ray hit your DRAM, or you have a memory corruption bug, you have bigger problems and should exit right aay.

Reply:
What if hypothetically I do:

```
for x in range(10):
    print(get_shader_string(x))
```

And then it just fucking dies?

Do you think that is good?

Reply:
Well Alem

Would you ever do that? Don't make strawman arguments. That's like saying "What if I dereference an arbitrary pointer? And then it just crashes? Is that a good idea?"

Think about the situations when this function would be called with an invalid input. What are these situations?

The only ones I can think of are:
- User memory corruption
- Bit flips from cosmic rays
- Malfunctioning hardware

Now, if any of those things happened, what do you think you should do? Do you think your program can keep running safely? Or should it exit ASAP and write a core dump so you can debug it?